src/FrameDump.cpp

```diff
@@ -0,0 +1,209 @@
+
+    // save these values here because its way easier this way.
+    ZDotDisp_Track = (GPU.GPU3D.ZeroDotWLimit != 0);
+    ZDotDisp = ((GPU.GPU3D.ZeroDotWLimit == 0) ? 0 : ((GPU.GPU3D.ZeroDotWLimit - 0x1FF) / 0x200));
+    PolyAttr = GPU.GPU3D.CurPolygonAttr;
+    PolyAttrUnset = GPU.GPU3D.PolygonAttr;
+    VtxColor = GPU.GPU3D.VertexColor[2] << 10 | GPU.GPU3D.VertexColor[1] << 5 | GPU.GPU3D.VertexColor[0];
+    Viewport = ((191 - GPU.GPU3D.Viewport[3]) & 0xFF) << 24 | GPU.GPU3D.Viewport[2] << 16 | ((191 - GPU.GPU3D.Viewport[1]) & 0xFF) << 8 | GPU.GPU3D.Viewport[0];
+    memcpy(ProjStack, GPU.GPU3D.ProjMatrixStack, sizeof(GPU.GPU3D.ProjMatrixStack));
+    memcpy(PosStack, GPU.GPU3D.PosMatrixStack, sizeof(GPU.GPU3D.PosMatrixStack));
+    memcpy(VecStack, GPU.GPU3D.VecMatrixStack, sizeof(GPU.GPU3D.VecMatrixStack));
+    memcpy(TexStack, GPU.GPU3D.TexMatrixStack, sizeof(GPU.GPU3D.TexMatrixStack));
+    memcpy(ProjMtx, GPU.GPU3D.ProjMatrix, sizeof(GPU.GPU3D.ProjMatrix));
+    memcpy(PosMtx, GPU.GPU3D.PosMatrix, sizeof(GPU.GPU3D.PosMatrix));
+    memcpy(VecMtx, GPU.GPU3D.VecMatrix, sizeof(GPU.GPU3D.VecMatrix));
+    memcpy(TexMtx, GPU.GPU3D.TexMatrix, sizeof(GPU.GPU3D.TexMatrix));
+    MatrixMode = GPU.GPU3D.MatrixMode;
+    Polygon = GPU.GPU3D.PolygonMode;
+    VtxXY = GPU.GPU3D.CurVertex[1] << 16 | GPU.GPU3D.CurVertex[0];
+    VtxZ = GPU.GPU3D.CurVertex[2];
+    TexCoord = GPU.GPU3D.TexCoords[1] << 16 | GPU.GPU3D.TexCoords[0]; // use final texcoords so i dont have to worry about setting up the matrix & params properly.
+    TexParam = GPU.GPU3D.TexParam;
+    TexPalette = GPU.GPU3D.TexPalette;
+    DiffAmbi = GPU.GPU3D.MatAmbient[2] << 26 | GPU.GPU3D.MatAmbient[1] << 21 | GPU.GPU3D.MatAmbient[0] << 16
+             | GPU.GPU3D.MatDiffuse[2] << 10 | GPU.GPU3D.MatDiffuse[1] << 5 | GPU.GPU3D.MatDiffuse[0];
+    SpecEmis = GPU.GPU3D.MatEmission[2] << 26 | GPU.GPU3D.MatEmission[1] << 21 | GPU.GPU3D.MatEmission[0] << 16
+             | GPU.GPU3D.MatSpecular[2] << 10 | GPU.GPU3D.MatSpecular[1] << 5 | GPU.GPU3D.MatSpecular[0];
+    for (int i = 0; i < 32; i++)
+    {
+        u8 tbl = i << 2;
+        Shininess[i] = GPU.GPU3D.ShininessTable[tbl];
+        Shininess[i] |= GPU.GPU3D.ShininessTable[tbl + 1] << 8;
+        Shininess[i] |= GPU.GPU3D.ShininessTable[tbl + 2] << 16;
+        Shininess[i] |= GPU.GPU3D.ShininessTable[tbl + 3] << 24;
+    }
+    for (u32 i = 0; i < 4; i++)
+    {
+        LightVec[i] = i << 30 | ((u32)GPU.GPU3D.LightDirection[i][2] & 0x3FF) << 20 | ((u32)GPU.GPU3D.LightDirection[i][1] & 0x3FF) << 10 | ((u32)GPU.GPU3D.LightDirection[i][0] & 0x3FF);
+        LightColor[i] = i << 30 | GPU.GPU3D.LightColor[i][2] << 10 | GPU.GPU3D.LightColor[i][1] << 5 | GPU.GPU3D.LightColor[i][0];
+    }
+    SwapBuffer = GPU.GPU3D.FlushAttributes;
+
+    // they should already be empty, but just in case.
+    Cmds.clear();
+    Params.clear();
+}
+
+bool FrameDump::FinFrameDump()
+{
+    std::string filename = ROMManager::GetFrameDumpName();
```

Comment:
This name should be passed in as an argument. The `FrameDump` doesn't need to know anything about the frontend, it just needs to be told where to write the file.